data/transactions/application.go

```diff
@@ -0,0 +1,776 @@
+			ad.EvalDelta = stateDeltas
+		} else {
+			// Ignore errors and rejections from the ClearStateProgram
+		}
+
+		// Fetch the (potentially updated) sender record
+		record, err = balances.Get(sender, false)
+		if err != nil {
+			ad.EvalDelta = basics.EvalDelta{}
+			return err
+		}
+	}
+
+	// Update the TotalAppSchema used for MinBalance calculation,
+	// since the sender no longer has to store LocalState
+	totalSchema := record.TotalAppSchema
+	localSchema := record.AppLocalStates[appIdx].Schema
+	totalSchema = totalSchema.SubSchema(localSchema)
+	record.TotalAppSchema = totalSchema
+
+	// Deallocate the AppLocalState and finish
+	record.AppLocalStates = cloneAppLocalStates(record.AppLocalStates)
+	delete(record.AppLocalStates, appIdx)
+
+	err = balances.Put(record)
+	if err != nil {
+		ad.EvalDelta = basics.EvalDelta{}
+	}
+	return err
+}
+
+func applyOptIn(balances Balances, sender basics.Address, appIdx basics.AppIndex, params basics.AppParams) error {
+	record, err := balances.Get(sender, false)
+	if err != nil {
+		return err
+	}
+
+	// If the user has already opted in, fail
+	_, ok := record.AppLocalStates[appIdx]
+	if ok {
+		return fmt.Errorf("account %s has already opted in to app %d", sender.String(), appIdx)
+	}
+
+	// Make sure the user isn't already at the app opt-in max
+	maxAppsOptedIn := balances.ConsensusParams().MaxAppsOptedIn
+	if len(record.AppLocalStates) >= maxAppsOptedIn {
+		return fmt.Errorf("cannot opt in app %d for %s: max opted-in apps per acct is %d", appIdx, sender.String(), maxAppsOptedIn)
+	}
+
+	// If the user hasn't opted in yet, allocate LocalState for the app
+	record.AppLocalStates = cloneAppLocalStates(record.AppLocalStates)
+	record.AppLocalStates[appIdx] = basics.AppLocalState{
+		Schema: params.LocalStateSchema,
+	}
+
+	// Update the TotalAppSchema used for MinBalance calculation,
+	// since the sender must now store LocalState
+	totalSchema := record.TotalAppSchema
+	totalSchema = totalSchema.AddSchema(params.LocalStateSchema)
+	record.TotalAppSchema = totalSchema
+
+	return balances.Put(record)
+}
+
+func (ac *ApplicationCallTxnFields) apply(header Header, balances Balances, spec SpecialAddresses, ad *ApplyData, txnCounter uint64, steva StateEvaluator) (err error) {
+	// Keep track of the application ID we're working on
+	appIdx := ac.ApplicationID
+
+	// this is not the case in the current code but still probably better to check
+	if ad == nil {
+		err = fmt.Errorf("cannot use empty ApplyData")
+		return
+	}
+
+	// Specifying an application ID of 0 indicates application creation
+	if ac.ApplicationID == 0 {
+		appIdx, err = ac.createApplication(balances, header.Sender, txnCounter)
+		if err != nil {
+			return
+		}
+	}
+
+	// Initialize our TEAL evaluation context. Internally, this manages
+	// access to balance records for Stateful TEAL programs. Stateful TEAL
+	// may only access
+	// - The sender's balance record
+	// - The balance records of accounts explicitly listed in ac.Accounts
```

Comment:
Can a TEAL program use this mechanism to read global in additional to local state from foreign applications (without specifying it them in `ForeignApps`)?